### PR TITLE
AArch64: Add vectorized String.hashCode() implementation

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
This commit adds vectorized String.hashCode() implementation to AArch64 code generator.

### CODE REVIEW

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -5950,6 +5951,263 @@ static TR::Register *inlineCRC32CUpdateBytes(TR::Node *node, TR::CodeGenerator *
...
+/**
+ * @brief Generates vectorized instruction sequence for computing hashcode of a String
+ *
+ * @param[in] node: node
+ * @param[in] isCompressed: true if compressed string
+ * @param[in] cg: CodeGenerator
+ *
+ * @returns register
+ */
+static TR::Register* inlineStringHashCode(TR::Node* node, bool isCompressed, TR::CodeGenerator* cg)
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:5963]: Please don't mix the `type* variable` style and the `type *variable` style.
The variables declared later in this function use the second style.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -5950,6 +5951,263 @@ static TR::Register *inlineCRC32CUpdateBytes(TR::Node *node, TR::CodeGenerator *
...
+   {
+   /*
+    *    str[16] = example string representing 16 compressed characters
+    *
+    *    The serial method for creating the hash:
+    *          hash = 0, offset = 0, count = 8
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:5969]: You mean `count = 16` ?

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -5950,6 +5951,263 @@ static TR::Register *inlineCRC32CUpdateBytes(TR::Node *node, TR::CodeGenerator *
...
+    *
+    *      (3) Vector[0] = Vector0[0] -> multiply 31^4 -> add Vector1[0] -> multiply 31^3
+    *          Vector[1] = Vector0[1] -> multiply 31^4 -> add Vector1[1] -> multiply 31^2
+    *          Vector[2] = Vector0[2] -> multiply 31^4 -> add Vector1[2] -> multiply 31^1
+    *          Vector[3] = Vector0[3] -> multiply 31^4 -> add Vector1[3] -> multiply 1
+    *
+    *    If the number of characters in the string is not a multiple of 16 (or 8 in the case of decompressed String),
+    *    then the remainder of the hash is calculated serially.
+    *
+    *	   addx dataAddrReg, addressReg, #headersize
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:6020]: The indentation of this line is different from that of other lines in the pseudo code.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -5950,6 +5951,263 @@ static TR::Register *inlineCRC32CUpdateBytes(TR::Node *node, TR::CodeGenerator *
...
+    *    b.gt residual
+    * done:
+    *    ret
+    *    .align 3
+    * multiplier:
+    *    .dword 0x3c10000745f   ; 31^2, 31^3
+    *    .dword 0x10000001f     ;    1, 31
+    * multiplier4:
+    *    .dword 0xe1781000e1781 ; 31^4
+    *    .dword 0xe1781000e1781
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:6086]: Maybe full digit representation such as `0x000003c10000745f` is better for these lines.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -5950,6 +5951,263 @@ static TR::Register *inlineCRC32CUpdateBytes(TR::Node *node, TR::CodeGenerator *
...
+      {
+      cg->generateDebugCounter(TR::DebugCounter::debugCounterName(comp, "cg.StringHashCode/(%s)/%s:long",
+                                                                        comp->signature(),
+                                                                        comp->getHotnessName()), *srm);
+      }
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::subimmw, node, lengthReg, lengthReg, (isCompressed ? 16 : 8));
+   TR::Register *multiplierReg = srm->findOrCreateScratchRegister(TR_VRF);
+   {
+      static uint32_t multiplier[] = { 31*31*31, 31*31, 31, 1 };
+      auto snippet = cg->findOrCreateConstantDataSnippet(node, multiplier, 16);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:6123]: Please add a comment for the magic number 16, or replace the number by `sizeof(multiplier)`, I guess.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fixed.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fixed.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fix the indentatino.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 4 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Changed them to full hex digit representation.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 5 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Replaced the number with `sizeof(multiplier)`.

**Comment by knn-k** [comment]
Jenkins compile alinux64,amac jdk11

**Comment by knn-k** [comment]
Jenkins compile alinux64,amac jdk17

**Review by knn-k — APPROVED** [review]
I ran the following tests with Java 17 on internal servers, and they all passed:
- sanity.functional on Linux: job/Grinder/36238/
- sanity.functional on macOS: job/Test_openjdk17_j9_sanity.functional_aarch64_mac_Personal/108/
- sanity.openjdk on macOS: job/Grinder/36239/

**Comment by tajila** [comment]
@Akira1Saitoh you'll need to double deliver to 0.43

**Pull request merged**